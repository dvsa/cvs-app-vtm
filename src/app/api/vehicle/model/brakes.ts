/**
 * Vehicles Microservice
 * This is the API spec for the vehicle microservice. Endpoints and parameters only exist for the operations getVehicle and getTechRecords. Other operations within the microservice are out of scope.
 *
 * OpenAPI spec version: 1.0.0
 * Contact: bpecete@deloittece.com
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */
import { BrakeForceWheelsNotLocked } from './brakeForceWheelsNotLocked';
import { BrakeForceWheelsUpToHalfLocked } from './brakeForceWheelsUpToHalfLocked';

export interface Brakes { 
    /**
     * Used only for PSV
     */
    brakeCodeOriginal?: string;
    /**
     * Used only for PSV
     */
    brakeCode?: string;
    /**
     * Used only for PSV
     */
    dataTrBrakeOne?: string;
    /**
     * Used only for PSV
     */
    dataTrBrakeTwo?: string;
    /**
     * Used only for PSV
     */
    dataTrBrakeThree?: string;
    /**
     * Used only for PSV
     */
    retarderBrakeOne?: Brakes.RetarderBrakeOneEnum;
    /**
     * Used only for PSV
     */
    retarderBrakeTwo?: Brakes.RetarderBrakeTwoEnum;
    /**
     * Used for PSV, HGV and TRL
     */
    dtpNumber?: string;
    brakeForceWheelsNotLocked?: BrakeForceWheelsNotLocked;
    brakeForceWheelsUpToHalfLocked?: BrakeForceWheelsUpToHalfLocked;
    /**
     * Used only for TRL
     */
    loadSensingValve?: boolean;
    /**
     * Used only for TRL
     */
    antilockBrakingSystem?: boolean;
}
export namespace Brakes {
    export type RetarderBrakeOneEnum = 'electric' | 'exhaust' | 'friction' | 'hydraulic' | 'other' | 'none';
    export const RetarderBrakeOneEnum = {
        Electric: 'electric' as RetarderBrakeOneEnum,
        Exhaust: 'exhaust' as RetarderBrakeOneEnum,
        Friction: 'friction' as RetarderBrakeOneEnum,
        Hydraulic: 'hydraulic' as RetarderBrakeOneEnum,
        Other: 'other' as RetarderBrakeOneEnum,
        None: 'none' as RetarderBrakeOneEnum
    };
    export type RetarderBrakeTwoEnum = 'electric' | 'exhaust' | 'friction' | 'hydraulic' | 'other' | 'none';
    export const RetarderBrakeTwoEnum = {
        Electric: 'electric' as RetarderBrakeTwoEnum,
        Exhaust: 'exhaust' as RetarderBrakeTwoEnum,
        Friction: 'friction' as RetarderBrakeTwoEnum,
        Hydraulic: 'hydraulic' as RetarderBrakeTwoEnum,
        Other: 'other' as RetarderBrakeTwoEnum,
        None: 'none' as RetarderBrakeTwoEnum
    };
}