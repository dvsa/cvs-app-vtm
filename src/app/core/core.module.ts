import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { GlobalErrorComponent } from './components/global-error/global-error.component';
import { SpinnerComponent } from './components/spinner/spinner.component';
import { FooterComponent } from './components/footer/footer.component';
import { HeaderComponent } from './components/header/header.component';
import { RouterModule } from '@angular/router';
import { BreadcrumbsComponent } from './components/breadcrumbs/breadcrumbs.component';
import { PageNotFoundComponent } from './components/page-not-found/page-not-found.component';
import { ServerErrorComponent } from './components/server-error/server-error.component';

@NgModule({
  declarations: [FooterComponent, HeaderComponent, GlobalErrorComponent, SpinnerComponent, BreadcrumbsComponent, PageNotFoundComponent, ServerErrorComponent],
  imports: [CommonModule, RouterModule],
  exports: [FooterComponent, HeaderComponent, GlobalErrorComponent, SpinnerComponent, BreadcrumbsComponent]
})
export class CoreModule {}
