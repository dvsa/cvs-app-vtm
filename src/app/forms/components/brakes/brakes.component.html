<ng-container [formGroup]="form">
  <table class="govuk-table">
    <thead class="govuk-table__head">
      <tr class="govuk-table__row">
        <th scope="col" class="govuk-table__header"></th>

        <th scope="col" class="govuk-table__header">Brake actuator</th>

        <th scope="col" class="govuk-table__header">Level length</th>

        <th scope="col" class="govuk-table__header">Spring brake parking</th>
      </tr>
    </thead>

    <tbody class="govuk-table__body" formArrayName="axles">
      <tr class="govuk-table__row" *ngFor="let axle of axles.controls; let i = index" [formGroupName]="i">
        <th scope="row" class="govuk-table__header">Axle {{ axle.value.axleNumber }}</th>

        <ng-container formGroupName="brakes">
          <td class="govuk-table__cell">
            <app-number-input
              formControlName="brakeActuator"
              [name]="'brakeActuator' + i"
              [isEditing]="isEditing"
              [noBottomMargin]="true"
            ></app-number-input>
          </td>

          <td class="govuk-table__cell">
            <app-number-input
              formControlName="leverLength"
              [name]="'leverLength' + i"
              [isEditing]="isEditing"
              [noBottomMargin]="true"
            ></app-number-input>
          </td>

          <td class="govuk-table__cell">
            <app-select
              formControlName="springBrakeParking"
              [name]="'springBrakeParking' + i"
              [options]="booleanOptions"
              [isEditing]="isEditing"
              [noBottomMargin]="true"
            ></app-select>
          </td>
        </ng-container>
      </tr>
    </tbody>
  </table>

  <div class="radio-buttons" formGroupName="brakes">
    <ng-template [ngTemplateOutlet]="radio" [ngTemplateOutletContext]="{ name: 'loadSensingValve' }"></ng-template>

    <ng-template [ngTemplateOutlet]="radio" [ngTemplateOutletContext]="{ name: 'antilockBrakingSystem' }"></ng-template>

    <ng-template #radio let-name="name">
      <app-radio-group
        [formControlName]="name"
        [name]="name"
        [label]="pascalCase(name)"
        [options]="booleanOptions"
        [isEditing]="isEditing"
      ></app-radio-group>
    </ng-template>
  </div>
</ng-container>
