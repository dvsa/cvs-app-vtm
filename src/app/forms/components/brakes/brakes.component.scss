.radio-buttons {
  display: flex;
  flex-flow: row wrap;
  justify-content: space-around;
}
