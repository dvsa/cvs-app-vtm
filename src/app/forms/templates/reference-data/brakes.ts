import { ValidatorNames } from '@forms/models/validators.enum';
import { FormNode, FormNodeEditTypes, FormNodeTypes } from '@forms/services/dynamic-form.types';

export const template: FormNode = {
  name: 'brakes',
  type: FormNodeTypes.GROUP,
  label: 'Brakes',
  children: [
    {
      name: 'resourceKey',
      label: 'Key',
      type: FormNodeTypes.CONTROL,
      editType: FormNodeEditTypes.TEXT,
      disabled: true,
      validators: [
        {
          name: ValidatorNames.Required
        }
      ]
    },
    {
      name: 'description',
      label: 'Description',
      type: FormNodeTypes.CONTROL,
      editType: FormNodeEditTypes.TEXT,
      validators: [
        {
          name: ValidatorNames.Required
        }
      ]
    },
    {
      name: 'service',
      label: 'Service',
      type: FormNodeTypes.CONTROL,
      editType: FormNodeEditTypes.TEXT,
      validators: [
        {
          name: ValidatorNames.MaxLength,
          args: 60
        }
      ]
    },
    {
      name: 'secondary',
      label: 'Secondary',
      type: FormNodeTypes.CONTROL,
      editType: FormNodeEditTypes.TEXT,
      validators: [
        {
          name: ValidatorNames.MaxLength,
          args: 60
        }
      ]
    },
    {
      name: 'parking',
      label: 'Parking',
      type: FormNodeTypes.CONTROL,
      editType: FormNodeEditTypes.TEXT,
      validators: [
        {
          name: ValidatorNames.MaxLength,
          args: 60
        }
      ]
    }
  ]
};

export const templateList = [
  {
    column: 'resourceKey',
    heading: 'Key',
    order: 1
  },
  {
    column: 'description',
    heading: 'Brakes',
    order: 2
  },
  {
    column: 'service',
    heading: 'Service',
    order: 3
  },
  {
    column: 'secondary',
    heading: 'Secondary',
    order: 4
  },
  {
    column: 'parking',
    heading: 'Parking',
    order: 5
  }
];
