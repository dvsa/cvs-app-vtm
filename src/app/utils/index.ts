export * from './mockStore';
export * from './test-model.utils';
export * from './testing.utils';
export * from './form.utils';


