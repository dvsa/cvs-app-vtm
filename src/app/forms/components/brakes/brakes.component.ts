import { Component, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { FormArray } from '@angular/forms';
import { DynamicFormService } from '@forms/services/dynamic-form.service';
import { CustomFormGroup, FormNode, FormNodeOption, FormNodeWidth } from '@forms/services/dynamic-form.types';
import { TrlBrakes } from '@forms/templates/trl/trl-brakes.template';
import { TechRecordModel } from '@models/vehicle-tech-record.model';
import { Subject, debounceTime, takeUntil } from 'rxjs';

@Component({
  selector: 'app-brakes[vehicleTechRecord]',
  templateUrl: './brakes.component.html',
  styleUrls: ['./brakes.component.scss']
})
export class BrakesComponent implements OnInit, OnDestroy {
  @Input() vehicleTechRecord!: TechRecordModel;
  @Input() isEditing = false;
  @Output() formChange = new EventEmitter();

  form!: CustomFormGroup;
  template!: FormNode;

  booleanOptions: Array<FormNodeOption<boolean>> = [{ value: true, label: 'Yes' }, { value: false, label: 'No' }];

  private destroy$ = new Subject<void>();

  constructor(private dfs: DynamicFormService) {}

  ngOnInit(): void {
    this.template = TrlBrakes;

    this.form = this.dfs.createForm(this.template, this.vehicleTechRecord) as CustomFormGroup;

    this.form.cleanValueChanges
      .pipe(debounceTime(400), takeUntil(this.destroy$))
      .subscribe(e => this.formChange.emit(e));
  }

  ngOnDestroy(): void {
    this.destroy$.next();
    this.destroy$.complete();
  }

  get axles(): FormArray {
    return this.form.get(['axles']) as FormArray;
  }
}
