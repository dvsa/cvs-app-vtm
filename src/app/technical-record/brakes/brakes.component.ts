import {
  Component,
  OnInit,
  ChangeDetectionStrategy,
  Input,
  OnChanges,
  SimpleChanges
} from '@angular/core';
import { Brakes, Axle, TechRecord } from '@app/models/tech-record.model';
import { VEHICLE_TYPES } from '@app/app.enums';

@Component({
  selector: 'vtm-brakes',
  templateUrl: './brakes.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class BrakesComponent implements OnChanges {
  @Input() techRecord: TechRecord;
  @Input() editState: boolean;

  brakes: Brakes;
  axles: Axle[];
  trlVvehicleType: VEHICLE_TYPES.TRL;
  psvVvehicleType: VEHICLE_TYPES.PSV;

  constructor() {}

  ngOnChanges(changes: SimpleChanges): void {
    const { techRecord } = changes;

    if (techRecord) {
      this.brakes = !!this.techRecord.brakes ? this.techRecord.brakes : ({} as Brakes);
      this.axles = this.techRecord.axles ? this.techRecord.axles : ([] as Axle[]);
    }
  }
}
