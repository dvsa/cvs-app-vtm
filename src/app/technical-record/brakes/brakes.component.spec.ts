import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { BrakesComponent } from './brakes.component';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { SharedModule } from '@app/shared/shared.module';
import { TESTING_UTILS } from '../../utils/testing.utils';

describe('BrakesComponent', () => {
  let component: BrakesComponent;
  let fixture: ComponentFixture<BrakesComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [SharedModule],
      declarations: [ BrakesComponent ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(BrakesComponent);
    component = fixture.componentInstance;
    component.currentVehicleType = 'trl';
    component.brakes = TESTING_UTILS.mockBrakes();
    fixture.detectChanges();
  });

  it('should create view only with populated data', () => {
    fixture.detectChanges();

    expect(component).toBeDefined();
    expect(fixture).toMatchSnapshot();
  });
});
