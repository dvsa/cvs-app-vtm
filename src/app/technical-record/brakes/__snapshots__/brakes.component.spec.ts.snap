// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrakesComponent should create view only with populated data 1`] = `
<vtm-brakes
  brakes={[Function Object]}
  currentVehicleType={[Function String]}
>
  <dl
    class="govuk-summary-list"
  >
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Load sensing valve
      </dt>
      <dd
        class="govuk-summary-list__value"
        id="test-brakes-loadSensingValve"
      >
         No 
      </dd>
      <dd
        class="govuk-summary-list__value"
      />
    </div>
    <div
      class="govuk-summary-list__row"
    >
      <dt
        class="govuk-summary-list__key"
      >
        Antilock braking system
      </dt>
      <dd
        class="govuk-summary-list__value"
        id="test-brakes-antilockBrakingSystem"
      >
         Yes 
      </dd>
      <dd
        class="govuk-summary-list__value"
      />
    </div>
  </dl>
</vtm-brakes>
`;
